src/extension/kernel.ts

```diff
@@ -1119,4 +1119,79 @@ export class Kernel implements Disposable {
   public getPlainCache(cacheId: string): Promise<Uint8Array> | undefined {
     return this.serializer?.getPlainCache(cacheId)
   }
+
+  async runProgram(program?: RunProgramOptions | string) {
+    let programOptions: RunProgramOptions
+
+    if (!this.runner) {
+      return Promise.reject(new Error('No runner available'))
+    }
+
+    if (typeof program === 'object') {
+      programOptions = program
+    } else if (typeof program === 'string') {
+      programOptions = {
+        programName: '',
+        background: false,
+        exec: {
+          type: 'script',
+          script: program,
+        },
+        languageId: 'sh',
+        storeLastOutput: false,
+        tty: true,
+      }
+    } else {
+      return Promise.reject(new Error('Invalid runProgram arguments'))
+    }
+
+    const programSession = await this.runner.createProgramSession(programOptions)
+    this.context.subscriptions.push(programSession)
+
+    let execRes: string | undefined
+    const onData = (data: string | Uint8Array) => {
+      if (execRes === undefined) {
+        execRes = ''
+      }
+      execRes += data.toString()
+    }
+
+    programSession.onDidWrite(onData)
+    programSession.onDidErr(onData)
+    await programSession.run()
+
+    return new Promise<{ exitReason?: RunnerExitReason; code?: number | void; output?: string }>(
+      (resolve, reject) => {
+        programSession.onDidClose(async (code) => {
+          return resolve({ code, output: execRes?.trim() })
+        })
+
+        programSession.onInternalErr((e) => {
+          reject(e)
+        })
+
+        const exitReason = programSession.hasExited()
+
+        if (exitReason) {
```

Comment:
When `exitReason` is null/undefined, which isn't impossible, the promise will dangle (never resolve or reject). The fix is likely to switch against `exitReason?.type` and make the default case error.

Reply:
Nvm. I just checked the "reference impl," and this code checks for premature termination of the program, so it's not expected always to be non-undefined. In fact, if it's defined it's almost always an error.